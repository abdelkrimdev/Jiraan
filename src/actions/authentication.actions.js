import { authenticationConstants } from '../constants/authentication.constants'
import { getCurrentUser, signIn, signOut, signUp } from '../services/authentication.service'

export const authenticationActions = {
  signUp: (email, password) => {
    return async (dispatch) => {
      try {
        dispatch(request())

        const payload = await signUp(email, password)

        dispatch(success(payload.user))
      } catch (error) {
        dispatch(failure(error))
      }
    }

    function request () {
      return { type: authenticationConstants.USER_CREATE_REQUEST }
    }
    function success (user) {
      return { type: authenticationConstants.USER_CREATE_SUCCESS, user }
    }
    function failure (error) {
      return { type: authenticationConstants.USER_CREATE_FAILURE, error }
    }
  },
  signIn: (email, password) => {
    return async (dispatch) => {
      try {
        dispatch(request())

        const payload = await signIn(email, password)

        dispatch(success(payload.user))
      } catch (error) {
        dispatch(failure(error))
      }
    }

    function request () {
      return { type: authenticationConstants.USER_LOGIN_REQUEST }
    }
    function success (user) {
      return { type: authenticationConstants.USER_LOGIN_SUCCESS, user }
    }
    function failure (error) {
      return { type: authenticationConstants.USER_LOGIN_FAILURE, error }
    }
  },
  getCurrentUser: () => {
    return (dispatch) => {
      getCurrentUser((user) => {
        dispatch(success(user))
      }, (error) => {
        dispatch(failure(error))
      })
    }

    function success (user) {
      return { type: authenticationConstants.USER_CURRENT_SUCCESS, user }
    }
    function failure (error) {
      return { type: authenticationConstants.USER_CURRENT_FAILURE, error }
    }
  },
  signOut: () => {
    return async (dispatch) => {
      try {
        dispatch(request())

        await signOut()

        dispatch(success())
      } catch (error) {
        dispatch(failure(error))
      }
    }

    function request () {
      return { type: authenticationConstants.USER_LOGOUT_REQUEST }
    }
    function success () {
      return { type: authenticationConstants.USER_LOGOUT_SUCCESS }
    }
    function failure (error) {
      return { type: authenticationConstants.USER_LOGOUT_FAILURE, error }
    }
  }
}
