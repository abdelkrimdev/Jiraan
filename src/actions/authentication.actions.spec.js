import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'

import { authenticationActions } from './authentication.actions'
import { getCurrentUser, signIn, signOut, signUp } from '../services/authentication.service'

jest.mock('../services/authentication.service')
const mockStore = configureStore([ thunkMiddleware ])

describe('authentication actions', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should create a user with email and password.', async () => {
    const store = mockStore()
    const email = 'user@somewhere.com'
    const password = 'password'

    signUp.mockResolvedValueOnce({
      user: {
        displayName: 'username',
        email: 'user@somewhere.com'
      }
    })

    await store.dispatch(authenticationActions.signUp(email, password))

    expect(signUp).toHaveBeenCalled()
    expect(signUp).toHaveBeenCalledWith(email, password)
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should capture sign-up errors.', async () => {
    const store = mockStore()
    const email = 'invalid-email.com'
    const password = 'weak-password'

    signUp.mockRejectedValueOnce({
      code: 'error/code',
      message: 'Create User Error Message'
    })

    await store.dispatch(authenticationActions.signUp(email, password))

    expect(signUp).toHaveBeenCalled()
    expect(signUp).toHaveBeenCalledWith(email, password)
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should login a user with email and password.', async () => {
    const store = mockStore()
    const email = 'user@somewhere.com'
    const password = 'password'

    signIn.mockResolvedValueOnce({
      user: {
        displayName: 'username',
        email: 'user@somewhere.com'
      }
    })

    await store.dispatch(authenticationActions.signIn(email, password))

    expect(signIn).toHaveBeenCalled()
    expect(signIn).toHaveBeenCalledWith(email, password)
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should capture sign-in errors.', async () => {
    const store = mockStore()
    const email = 'user@somewhere.com'
    const password = 'wrong-password'

    signIn.mockRejectedValueOnce({
      code: 'error/code',
      message: 'Login User Error Message'
    })

    await store.dispatch(authenticationActions.signIn(email, password))

    expect(signIn).toHaveBeenCalled()
    expect(signIn).toHaveBeenCalledWith(email, password)
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should get the current user.', () => {
    const store = mockStore()

    getCurrentUser.mockImplementation((success, failure) => {
      const currentUser = {
        displayName: 'username',
        email: 'user@somewhere.com'
      }

      success(currentUser)
    })

    store.dispatch(authenticationActions.getCurrentUser())

    expect(getCurrentUser).toHaveBeenCalled()
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should capture get-current-user errors.', () => {
    const store = mockStore()

    getCurrentUser.mockImplementation((success, failure) => {
      const error = {
        code: 'error/code',
        message: 'Get Current User Error Message'
      }

      failure(error)
    })

    store.dispatch(authenticationActions.getCurrentUser())

    expect(getCurrentUser).toHaveBeenCalled()
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should logout the user.', async () => {
    const store = mockStore()

    signOut.mockResolvedValueOnce()

    await store.dispatch(authenticationActions.signOut())

    expect(signOut).toHaveBeenCalled()
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should capture sign-out errors.', async () => {
    const store = mockStore()

    signOut.mockRejectedValueOnce({
      code: 'error/code',
      message: 'Sign Out User Error Message'
    })

    await store.dispatch(authenticationActions.signOut())

    expect(signOut).toHaveBeenCalled()
    expect(store.getActions()).toMatchSnapshot()
  })
})
