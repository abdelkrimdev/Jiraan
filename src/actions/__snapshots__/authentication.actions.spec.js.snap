// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authentication actions should capture get-current-user errors. 1`] = `
Array [
  Object {
    "error": Object {
      "code": "error/code",
      "message": "Get Current User Error Message",
    },
    "type": "FETCH_GET_CURRENT_USER_FAILURE",
  },
]
`;

exports[`authentication actions should capture sign-in errors. 1`] = `
Array [
  Object {
    "type": "FETCH_LOGIN_USER_REQUEST",
  },
  Object {
    "error": Object {
      "code": "error/code",
      "message": "Login User Error Message",
    },
    "type": "FETCH_LOGIN_USER_FAILURE",
  },
]
`;

exports[`authentication actions should capture sign-out errors. 1`] = `
Array [
  Object {
    "type": "FETCH_LOGOUT_USER_REQUEST",
  },
  Object {
    "error": Object {
      "code": "error/code",
      "message": "Sign Out User Error Message",
    },
    "type": "FETCH_LOGOUT_USER_FAILURE",
  },
]
`;

exports[`authentication actions should capture sign-up errors. 1`] = `
Array [
  Object {
    "type": "FETCH_CREATE_USER_REQUEST",
  },
  Object {
    "error": Object {
      "code": "error/code",
      "message": "Create User Error Message",
    },
    "type": "FETCH_CREATE_USER_FAILURE",
  },
]
`;

exports[`authentication actions should create a user with email and password. 1`] = `
Array [
  Object {
    "type": "FETCH_CREATE_USER_REQUEST",
  },
  Object {
    "type": "FETCH_CREATE_USER_SUCCESS",
    "user": Object {
      "displayName": "username",
      "email": "user@somewhere.com",
    },
  },
]
`;

exports[`authentication actions should get the current user. 1`] = `
Array [
  Object {
    "type": "FETCH_GET_CURRENT_USER_SUCCESS",
    "user": Object {
      "displayName": "username",
      "email": "user@somewhere.com",
    },
  },
]
`;

exports[`authentication actions should login a user with email and password. 1`] = `
Array [
  Object {
    "type": "FETCH_LOGIN_USER_REQUEST",
  },
  Object {
    "type": "FETCH_LOGIN_USER_SUCCESS",
    "user": Object {
      "displayName": "username",
      "email": "user@somewhere.com",
    },
  },
]
`;

exports[`authentication actions should logout the user. 1`] = `
Array [
  Object {
    "type": "FETCH_LOGOUT_USER_REQUEST",
  },
  Object {
    "type": "FETCH_LOGOUT_USER_SUCCESS",
  },
]
`;
